pkg/upgrade/upgrade_state.go

```diff
@@ -0,0 +1,385 @@
+	}
+
+	m.Log.V(consts.LogLevelInfo).Info("Upgrades in progress",
+		"currently in progress", upgradesInProgress,
+		"max parallel upgrades", upgradePolicy.MaxParallelUpgrades,
+		"upgrade slots available", upgradesAvailable)
+
+	// First, check if unknown or ready nodes need to be upgraded
+	err := m.ProcessDoneOrUnknownNodes(ctx, currentState, UpgradeStateUnknown)
+	if err != nil {
+		m.Log.V(consts.LogLevelError).Error(err, "Failed to process nodes", "state", UpgradeStateUnknown)
+		return err
+	}
+	err = m.ProcessDoneOrUnknownNodes(ctx, currentState, UpgradeStateDone)
+	if err != nil {
+		m.Log.V(consts.LogLevelError).Error(err, "Failed to process nodes", "state", UpgradeStateDone)
+		return err
+	}
+	// Start upgrade process for upgradesAvailable number of nodes
+	err = m.ProcessUpgradeRequiredNodes(ctx, currentState, upgradesAvailable)
+	if err != nil {
+		m.Log.V(consts.LogLevelError).Error(
+			err, "Failed to process nodes", "state", UpgradeStateUpgradeRequired)
+		return err
+	}
+	// Schedule nodes for drain
+	err = m.ProcessDrainNodes(ctx, currentState, upgradePolicy.DrainSpec)
+	if err != nil {
+		m.Log.V(consts.LogLevelError).Error(err, "Failed to schedule nodes drain")
+		return err
+	}
+	err = m.ProcessPodRestartNodes(ctx, currentState)
+	if err != nil {
+		m.Log.V(consts.LogLevelError).Error(err, "Failed to schedule pods restart")
+		return err
+	}
+	err = m.ProcessDrainFailedNodes(ctx, currentState)
+	if err != nil {
+		m.Log.V(consts.LogLevelError).Error(err, "Failed to process nodes which failed to drain")
+		return err
+	}
+	err = m.ProcessUncordonRequiredNodes(ctx, currentState)
+	if err != nil {
+		m.Log.V(consts.LogLevelError).Error(err, "Failed to uncordon nodes")
+		return err
+	}
+	m.Log.V(consts.LogLevelInfo).Info("State Manager, finished processing")
+	return nil
+}
+
+// ProcessDoneOrUnknownNodes iterates over UpgradeStateDone or UpgradeStateUnknown nodes and determines
+// whether each specific node should be in UpgradeStateUpgradeRequired or UpgradeStateDone state.
+func (m *ClusterUpgradeStateManager) ProcessDoneOrUnknownNodes(
+	ctx context.Context, currentClusterState *ClusterUpgradeState, nodeStateName string) error {
+	m.Log.V(consts.LogLevelInfo).Info("ProcessDoneOrUnknownNodes")
+
+	for _, nodeState := range currentClusterState.NodeStates[nodeStateName] {
+		podTemplateGeneration, err := utils.GetPodTemplateGeneration(nodeState.DriverPod, m.Log)
+		if err != nil {
+			m.Log.V(consts.LogLevelError).Error(
+				err, "Failed to get pod template generation", "pod", nodeState.DriverPod)
+			return err
+		}
+		if podTemplateGeneration != nodeState.DriverDaemonSet.GetGeneration() {
+			err := m.NodeUpgradeStateProvider.ChangeNodeUpgradeState(ctx, nodeState.Node, UpgradeStateUpgradeRequired)
+			if err != nil {
+				m.Log.V(consts.LogLevelError).Error(
+					err, "Failed to change node upgrade state", "state", UpgradeStateUpgradeRequired)
+				return err
+			}
+			return nil
```

Comment:
you are returning on the first  node which has its mofed pod generation different than the daemonset

is that the desired behavior ? i think you would want to process all nodes in the specified state.

Reply:
No, thanks for noticing. Replaced the `return` with `continue`